src/admin.c

```diff
@@ -318,77 +411,90 @@ main(int argc, char** argv)
 }
 
 static int
-master_key(char* password, bool generate_pwd, int pwd_length)
+master_key(char* password, bool generate_pwd, int pwd_length, char* filename)
 {
    FILE* file = NULL;
    char buf[MISC_LENGTH];
    char* encoded = NULL;
    struct stat st = {0};
    bool do_free = true;
-
-   if (pgagroal_get_home_directory() == NULL)
+   if (filename == NULL)
    {
-      char* username = pgagroal_get_user_name();
+      if (pgagroal_get_home_directory() == NULL)
+      {
+         char* username = pgagroal_get_user_name();
+
+         if (username != NULL)
+         {
+            warnx("No home directory for user \'%s\'", username);
+         }
+         else
+         {
+            warnx("No home directory for user running pgagroal");
+         }
 
-      if (username != NULL)
+         goto error;
+      }
+
+      memset(&buf, 0, sizeof(buf));
+      snprintf(&buf[0], sizeof(buf), "%s/.pgagroal", pgagroal_get_home_directory());
+
+      if (stat(&buf[0], &st) == -1)
       {
-         warnx("No home directory for user \'%s\'", username);
+         mkdir(&buf[0], S_IRWXU);
```

Comment:
should we catch a failure here?

Reply:
I haven't made any modifications to this part. This is already written in the master branch. I just indented it for those cases where filename is not provided.
I can look into it if there is some error?

Reply:
> I haven't made any modifications to this part. This is already written in the master branch. I just indented it for those cases where filename is not provided. I can look into it if there is some error?

Ok, if you will, please check for possible errors, or will postpone.